From 72360ef72bba88a557eb38c110d1ff7244eb2baa Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Tue, 7 Oct 2014 17:53:32 -0400
Subject: [md] raid1: fix_read_error should act on all non-faulty devices

Message-id: <1412704413-32535-2-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 97132
O-Subject: [RHEL7.1 PATCH 1/2] md/raid1: fix_read_error should act on all non-faulty devices.
Bugzilla: 1149931
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: NeilBrown <neilb@suse.de>

If a devices is being recovered it is not InSync and is not Faulty.

If a read error is experienced on that device, fix_read_error()
will be called, but it ignores non-InSync devices.  So it will
neither fix the error nor fail the device.

It is incorrect that fix_read_error() ignores non-InSync devices.
It should only ignore Faulty devices.  So fix it.

This became a bug when we allowed reading from a device that was being
recovered.  It is suitable for any subsequent -stable kernel.

Fixes: da8840a747c0dbf49506ec906757a6b87b9741e9
Cc: stable@vger.kernel.org (v3.5+)
Reported-by: Alexander Lyakas <alex.bolshoy@gmail.com>
Tested-by: Alexander Lyakas <alex.bolshoy@gmail.com>
Signed-off-by: NeilBrown <neilb@suse.de>
(cherry picked from commit b8cb6b4c121e1bf1963c16ed69e7adcb1bc301cd)
Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/raid1.c b/drivers/md/raid1.c
index 3966bd8..8245ac0 100644
--- a/drivers/md/raid1.c
+++ b/drivers/md/raid1.c
@@ -2051,7 +2051,7 @@ static void fix_read_error(struct r1conf *conf, int read_disk,
    d--;
    rdev = conf->mirrors[d].rdev;
    if (rdev &&
-       test_bit(In_sync, &rdev->flags))
+       !test_bit(Faulty, &rdev->flags))
     r1_sync_page_io(rdev, sect, s,
       conf->tmppage, WRITE);
   }
@@ -2063,7 +2063,7 @@ static void fix_read_error(struct r1conf *conf, int read_disk,
    d--;
    rdev = conf->mirrors[d].rdev;
    if (rdev &&
-       test_bit(In_sync, &rdev->flags)) {
+       !test_bit(Faulty, &rdev->flags)) {
     if (r1_sync_page_io(rdev, sect, s,
           conf->tmppage, READ)) {
      atomic_add(s, &rdev->corrected_errors);
-- 
1.7.1